framework/src/bcs/NodalBC.C

```diff
@@ -47,7 +47,60 @@ NodalBC::computeResidual(NumericVector<Number> & residual)
 }
 
 void
-NodalBC::computeJacobian(SparseMatrix<Number> & /*jacobian*/)
+NodalBC::computeJacobian()
 {
-  mooseError("This shouldn't be called!");
+  // We call the user's computeQpJacobian() function and store the
+  // results in the _assembly object. We can't store them directly in
+  // the element stiffness matrix, as they will only be inserted after
+  // all the assembly is done.
+  if (_var.isNodalDefined())
+  {
+    _qp = 0;
+    Real cached_val = computeQpJacobian();
+    dof_id_type cached_row = _var.nodalDofIndex();
+
+    // We'll cache the users computeQpJacobian() value in the Assembly object.
+    Assembly::CachedNodalBCData & cached_data = _fe_problem.assembly(0).cachedNodalBCData();
+
+    // Cache (i,j,value)
+    cached_data.cached_nodal_bc_vals.push_back(cached_val);
+    cached_data.cached_nodal_bc_rows.push_back(cached_row);
+    cached_data.cached_nodal_bc_cols.push_back(cached_row);
+  }
+}
+
+void
+NodalBC::computeOffDiagJacobian(unsigned int jvar)
+{
+  if (jvar == _var.number())
+    computeJacobian();
+  else
+  {
+    _qp = 0;
+    Real cached_val = computeQpOffDiagJacobian(jvar);
+    dof_id_type cached_row = _var.nodalDofIndex();
+    // Note: this only works for Lagrange variables...
+    dof_id_type cached_col = _current_node->dof_number(_sys.number(), jvar, 0);
+
+    // We'll cache the users computeQpJacobian() value in the Assembly object.
+    Assembly::CachedNodalBCData & cached_data = _assembly.cachedNodalBCData();
+
+    // Cache (i,j,value)
+    cached_data.cached_nodal_bc_vals.push_back(cached_val);
```

Comment:
Scary: Make sure these vectors have enough `capacity()` to keep re-allocs from happening (the first time it's probably unavoidable).